scripts/codex_autoreview.py

```diff
@@ -0,0 +1,641 @@
+    pr_evt = event["pull_request"]
+    pr_number = int(pr_evt["number"]) if "number" in pr_evt else int(event.get("number", 0))
+    # Fetch fresh PR to ensure up-to-date head SHA for comment anchoring
+    pr = _get_pr(owner, repo, pr_number, token)
+    head_sha = pr.get("head", {}).get("sha")
+    if not head_sha:
+        print("Missing head commit SHA", file=sys.stderr)
+        sys.exit(1)
+
+    changed = _list_changed_files(owner, repo, pr_number, token)
+
+    # Load guidelines and compose the single large prompt
+    guidelines = _load_guidelines()
+    repo_root = Path(os.environ.get("GITHUB_WORKSPACE", ".")).resolve()
+    prompt = _compose_prompt(guidelines, repo_root, changed, pr)
+
+    _dbg(1, f"Repo: {owner}/{repo} PR: #{pr_number} head={head_sha}")
+    _dbg(1, f"Changed files: {len(changed)}")
+    for fch in changed[:10]:  # cap to avoid log spam
+        _dbg(
+            2,
+            f" - {fch.filename} status={fch.status} patch_len={len(fch.patch.splitlines()) if fch.patch else 0}",
+        )
+
+    _dbg(2, f"Prompt length: {len(prompt)} chars")
+    print("Running Codex to generate review findings...", flush=True)
+    try:
+        output = _run_codex(prompt)
+    except Exception as e:  # pragma: no cover - robust reporting
+        print(f"Codex execution failed: {e}", file=sys.stderr)
+        sys.exit(2)
+
+    # Parse JSON output
+    try:
+        result = json.loads(output)
+    except json.JSONDecodeError as e:
+        print("Model did not return valid JSON:")
+        print(output)
+        print(f"JSON error: {e}", file=sys.stderr)
+        sys.exit(3)
+
+    findings: list[dict[str, Any]] = list(result.get("findings", []) or [])
+    overall = str(result.get("overall_correctness", "")).strip() or "patch is correct"
+    overall_explanation = str(result.get("overall_explanation", "")).strip()
+    overall_conf = result.get("overall_confidence_score")
+
+    summary_lines = [
+        "Codex Autonomous Review:",
+        f"- Overall: {overall}",
+        f"- Findings: {len(findings)}",
+    ]
+    if overall_explanation:
+        summary_lines.append("")
+        summary_lines.append(overall_explanation)
+    if isinstance(overall_conf, int | float):
+        summary_lines.append(f"Confidence: {overall_conf}")
+    summary = "\n".join(summary_lines)
+
+    try:
+        _post_review(owner=owner, repo=repo, pr_number=pr_number, token=token, summary=summary)
+    except HTTPError as e:  # pragma: no cover - external API
+        print(f"Failed to post review summary: {e}", file=sys.stderr)
+
+    # Build fast lookup of valid commentable lines per changed file
+    valid_lines_by_path: dict[str, set[int]] = {}
+    position_by_path: dict[str, dict[int, int]] = {}
+    for fch in changed:
+        if fch.patch:
+            valid_lines_by_path[fch.filename] = _parse_valid_head_lines_from_patch(fch.patch)
+            # Precompute positions for valid head lines
+            pos_map: dict[int, int] = {}
```

Comment:
[P1] Replace union in isinstance with tuple to avoid TypeError

In scripts/codex_autoreview.py (main()), the check `isinstance(overall_conf, int | float)` will raise `TypeError: isinstance() argument 2 cannot be a union` on Python 3.10+ because `isinstance` does not accept PEP 604 unions. This prevents the workflow from posting the review summary for any run that reaches this branch. Replace the union with a tuple of classes.
```suggestion
    if isinstance(overall_conf, (int, float)):
```